CPP client compilation issues.

**Describe the bug**

When enabling `-Wall` and `-Werror`, the library cannot be successfully compiled:

1. Non-virtual destructor in base class: `pulsar-client-cpp/lib/auth/AuthOauth2.h:47:7: error: base class 'class pulsar::CachedToken' has accessible non-virtual destructor [-Werror=non-virtual-dtor]`.
2. Unnecessary NULL checking on function pointer: `/pulsar-client-cpp/lib/ClientImpl.cc:100:10: error: the address of 'pulsar::Logger* logger()' will never be NULL [-Werror=address]`.

**To Reproduce**

Compile with `-Wall` and `-Werror`.

**Desktop (please complete the following information):**
 - OS: Linux
 - Compiler: GCC 7.3

--- a/pulsar-client-cpp/include/pulsar/Authentication.h
+++ b/pulsar-client-cpp/include/pulsar/Authentication.h
@@ -229,7 +229,7 @@ typedef std::shared_ptr<Oauth2Flow> FlowPtr;
 
 class CachedToken {
    public:
-    ~CachedToken();
+    virtual ~CachedToken();
     virtual bool isExpired() = 0;
     virtual AuthenticationDataPtr getAuthData() = 0;
 

--- a/pulsar-client-cpp/lib/ClientImpl.cc
+++ b/pulsar-client-cpp/lib/ClientImpl.cc
@@ -97,20 +97,18 @@ ClientImpl::ClientImpl(const std::string& serviceUrl, const ClientConfiguration&
       requestIdGenerator_(0),
       closingError(ResultOk) {
     std::unique_ptr<LoggerFactory> loggerFactory = clientConfiguration_.impl_->takeLogger();
-    if (!logger) {
 #ifdef USE_LOG4CXX
-        if (!clientConfiguration_.getLogConfFilePath().empty()) {
-            // A log4cxx log file was passed through deprecated parameter. Use that to configure Log4CXX
-            loggerFactory = Log4CxxLoggerFactory::create(clientConfiguration_.getLogConfFilePath());
-        } else {
-            // Use default simple console logger
-            loggerFactory = SimpleLoggerFactory::create();
-        }
-#else
+    if (!clientConfiguration_.getLogConfFilePath().empty()) {
+        // A log4cxx log file was passed through deprecated parameter. Use that to configure Log4CXX
+        loggerFactory = Log4CxxLoggerFactory::create(clientConfiguration_.getLogConfFilePath());
+    } else {
         // Use default simple console logger
         loggerFactory = SimpleLoggerFactory::create();
-#endif
     }
+#else
+    // Use default simple console logger
+    loggerFactory = SimpleLoggerFactory::create();
+#endif
     LogUtils::setLoggerFactory(std::move(loggerFactory));
 
     if (serviceUrl_.compare(0, 4, "http") == 0) {

